Move the non-IO parts of ceylon.io into ceylon.buffer

ASzc opened the pull request:
Fixes #449 (see that for more details) by moving the relevant code into `ceylon.buffer`, adding a ByteBuffer implementation for JS, and updating all the imports in the SDK. I would like to make further changes to the code, but that's for another PR.

The tests of each modified module pass. I have pushed my workaround for ceylon/ceylon.language#794, so the `ceylon.buffer` tests pass on JS. This may be a performance reduction, and can be reverted later once copyTo is fixed.

ASzc commented:
@FroMage could you review this please? Thanks.

gavinking commented:
So frankly I'm still unhappy with this stuff. The criticisms I made a couple of years ago still have not been addressed. So my question are:

- What is the purpose of `Buffer<T>`? Do we really have, or expect to have, code which abstracts over character buffers and byte buffers? If not, then coding to `Buffer` is just a useless thing and can never possibly perform acceptably due to boxing.
- Why is the only implementation of `CharacterBuffer` read-only? If there are really truly read-only buffers, then why isn't there a `ReadOnlyBuffer` interface instead of exception-throwing methods. And if so, how is a `ReadOnlyBuffer` different to a `List`? If _not_, then why don't we have a proper implementation of `CharacterBuffer`?
- Why isn't it possible to iterate buffers efficiently using `each()`. Shouldn't that be the programming model we're promoting here, not `iterator()` and `get()`, which are each problematic in different ways?

ASzc commented:
@gavinking I saw that issue (#267), and was planning on looking at that myself. I intended this PR to just be the separation + minimum changes to make it work, and do actual changes later. If a wider scope is desired now, then I will address those points here.

gavinking commented:
OK, sorry.

ASzc commented:
I'm not sure what the timeframe is for 1.2.1, but if it's going to be less than a few weeks, then maybe it is best to do all the changes together, so if it's not ready 1.2.1 can go out cleanly without it?

gavinking commented:
Right, I would like to see all this mess cleaned up for 1.2.1. It's been stewing for much too long already.

gavinking commented:
@ASzc well actually I think I lied. As of today even `each()` requires boxing :-(

That's something we really need to fix, but we just didn't get the bandwidth to do it yet.

ASzc commented:
@gavinking What was it that you were thinking prevented efficient use of each in ByteBuffer in your initial comment? Are there any special things I need to do, or is using the inherited `Iterable.each` sufficient?

FroMage commented:
Just added you to the SDK team!

ASzc commented:
I'm currently blocked by ceylon/ceylon-ide-eclipse#1629, so I can't write tests to solidify all this yet, but generally I have:

- Redesigned `Charset` and `Base64` as a unified `Codec` system, with a symmetrical encode/decode API. I plan to eventually add things like `Gzip` and `Base32` under the same type structure.
- Kept `Buffer<T>`, as it's useful as a type parameter constraint.
- Added `Buffer.array` for accessing the underlying array, and added constructors for `{T*}`, `Array<T>`, and `Integer`.
- Made `CharacterBuffer` conform to the `Buffer` interface (it's a proper read/write buffer now).

Questions:

- The split between `ByteBuffer` and `CharacterBuffer` is problematic. If we just had `Buffer<T>`, that would eliminate code duplication in the `Buffer` types, and the implementation hoop-jumping in the `Codec` types (see [codecs.ceylon](https://github.com/ASzc/ceylon-sdk/blob/io-buffer-split/source/ceylon/buffer/codec/codecs.ceylon) for example). I have left them split to satisfy these requirements for the JVM implementation:
  - Conversion to a JavaByteBuffer should involve no/minimal allocations.
  - The JavaByteBuffer should remain mutually in sync with the Ceylon ByteBuffer, when changes are made from either interface.
- I don't know type parameters and the `native` annotation well enough (`native` seems poorly documented honestly), but it seems I can't have a standard pure-Ceylon `Buffer<T>` for non-JVM use, and then if `T` satisfies `Byte`, *and* we're on the JVM, switch efficiently to being a `JavaByteBuffer` wrapper.
- I've been running into many situations with this where I wish I had more powerful type parameters, is any of this available or planned?
  - If type parameter `A` satisfies type `Qwe`, then define type parameter `B` with constraints x, y and z. Otherwise parameter `B` satisfies `Asd`.
  - Related to that, but different, the ability to switch on type parameters (I don't always have a instance of the thing to use `is` on!), described [here](http://ceylon-lang.org/documentation/1.2/spec/html_single/index.html#generictypeconstraints) but not implemented.

FroMage commented:
That sounds great, why not merge this?

ASzc commented:
I prefer not to merge things unless I have some certainty they work. I will be looking at this again in the next few days, maybe the blocker bug has fixed itself while I was away these past two weeks.

ASzc commented:
Poked at the bug again, no change but some new information was discovered. See ceylon/ceylon#5892.

quintesse commented:
ok to test

quintesse commented:
ok to test
test this please

quintesse commented:
ok to test
test this please
@quintesse kicks the CI server

quintesse commented:
The code doesn't compile:

```
compile-jvm:
[ceylon-compile] /home/ceylon/ceylon/ceylon-sdk/source/ceylon/io/module.ceylon:10: error: cannot find module artifact ceylon.buffer-1.2.1(.car|.jar)
[ceylon-compile]     shared import ceylon.buffer "1.2.1";
[ceylon-compile]     ^
[ceylon-compile]   	- dependency tree: ceylon.io/1.2.1 -> ceylon.buffer/1.2.1
[ceylon-compile] Note: Created module ceylon.test/1.2.1
[ceylon-compile] Note: Created module ceylon.locale/1.2.1
[ceylon-compile] ceylon compile: There was 1 error
```

ASzc commented:
I added `ceylon.buffer` to the `moduleset`s in build.xml. Is it working like it should now @quintesse? Jenkins reports a success now, but there are no test results, and I see "ceylon.buffer: command not found" in the build log.

ASzc commented:
I think I have it in a good state now. I will merge tomorrow if no issues arise.

I'm not able to run the tests on JS for some reason (Node seems to hang when running the first test), but I haven't made any modifications to `native` code since I last did.

ASzc commented:
It seems the release of 1.2.1 is too close, so I will wait until that's out to merge.

jvasileff commented on source/ceylon/buffer/Buffer.ceylon:
```diff
@@ -0,0 +1,172 @@
+    size => available;
+    
+    shared actual default Boolean equals(Object that) {
+        if (is Buffer<Element> that) {
```
Should this have `if (this === that) { return true; }` ?

jvasileff commented on source/ceylon/buffer/Buffer.ceylon:
```diff
@@ -0,0 +1,172 @@
+ you do with [[put]] until you reach the [[limit]].
+ 
+ Then if you want to read the data you just put in the
+ buffer, you [[flip]] the buffer, which will and set its
```
which will ~~and~~ set its

jvasileff commented on source/ceylon/buffer/Buffer.ceylon:
```diff
@@ -0,0 +1,172 @@
+ You can then start calling [[get]] to read objects from
+ this buffer until you reach its [[limit]].
+ 
+ Once you are done reading from the buffer, you will
```
s/will/can

ePaul commented on source/ceylon/buffer/Buffer.ceylon:
```diff
@@ -0,0 +1,180 @@
+     Reallocations caused by a call to [[resize]] will cause any
+     prior references obtained though this attribute to become
+     stale."
+    shared formal Array<Element> array;
```
Doesn't this limit implementations too much? I.e. this only allows implementations based on arrays, not on anything else, like mapped memory or similar.

ASzc replied:
It's important for symmetry in the `Codec` system that there's an equivalent to `String` to return from Byte-form encode/decode operations. This is the only way of producing that without copying.

I don't see `array` preventing a mapped file implementation, all that such a implementation would need to do is allocate an `Array` on-demand with a copy of the actual backing data. This isn't efficient obviously, but it wouldn't make sense to use such a buffer in the same way as `Array`-based `Buffers` are.

ASzc commented:
Rebased, and three extra commits added.

The JavaScript `ceylon.buffer` tests still hang. I was hoping it would be fixed in the 1.2.1 release, but no luck with a fresh build from a few hours ago. I've inserted print statements to try and trace the execution, but it just doesn't proceed any further on the call to `base16Byte.decode`. I'm going to file a bug, probably should wait until that is resolved before merging this.

ASzc commented:
The bug was fixed, all tests pass. If there are no objections, I will merge this evening or tomorrow.

quintesse commented:
@ASzc the problem, as I mentioned on the bug issue, is that because of this bug the module can't be released this way in the 1.2.1 timeframe. You either have to wait until 1.2.2 to publish it on the Herd or try to find a workaround for the bug.

quintesse commented:
Btw, not saying you can't merge, but if you had hoped to see it published on the Herd soon it won't be possible.

ASzc commented:
@quintesse the 1.2.1 SDK modules have been released as well, no? This PR is probably too impactful to include in a maintenance release for 1.2.1. My understanding is that changes on master are now destined for 1.2.2 (which seems like an appropriate settling period to me), is that correct?

quintesse commented:
Well the SDK release cycle is somewhat independent from the distribution cycle. And will only become more so in the future. So new modules can be added at any time provided they work with the currently released compiler. But if you're happy to wait for 1.2.2 then there's no problem :)
